Bottom: 6d478f0b18d54c3d92fa925462ba680378677de3
Top:    aeef3e58f6c29bfbf2e6d27c344d5a2d90d748b2
Author: Robert Collins <robertc@robertcollins.net>
Date:   2013-02-17 16:12:38 +1300

Add a dedicated interface for controlling an in-progress TestRun.


---

diff --git a/NEWS b/NEWS
index c298aa6..4c300de 100644
--- a/NEWS
+++ b/NEWS
@@ -35,6 +35,12 @@ Improvements
   series of dicts describing a test. Useful for writing trivial stream
   analysers. (Robert Collins)
 
+* New support class ``TestControl`` which permits cancelling an in-progress
+  run. (Robert Collins)
+
+* New support class ``StreamFailFast`` which calls a ``TestControl`` instance
+  to abort the test run when a failure is detected. (Robert Collins)
+
 * New ``TestCase`` decorator ``DecorateTestCaseResult`` that adapts the
   ``TestResult`` or ``StreamResult`` a case will be run with, for ensuring that
   a particular result object is used even if the runner running the test doesn't
diff --git a/doc/for-framework-folk.rst b/doc/for-framework-folk.rst
index 0f0537d..7162e97 100644
--- a/doc/for-framework-folk.rst
+++ b/doc/for-framework-folk.rst
@@ -174,7 +174,11 @@ TestControl
 
 Lastly we define the ``TestControl`` API which is used to provide the
 ``shouldStop`` and ``stop`` elements from ``TestResult``. Again, see the API
-documentation for ``testtools.TestControl``.
+documentation for ``testtools.TestControl``. ``TestControl`` can be paired with
+a ``StreamFailFast`` to trigger failing aborting a test run when a failure
+is observed. Aborting multiple workers in a distributed environment requires
+hooking whatever signalling mechanism the distributed environment has up to
+a ``TestControl`` in each worker process.
 
 StreamToDict
 ------------
diff --git a/testtools/__init__.py b/testtools/__init__.py
index a304ade..8b45e5e 100644
--- a/testtools/__init__.py
+++ b/testtools/__init__.py
@@ -27,9 +27,11 @@ __all__ = [
     'skip',
     'skipIf',
     'skipUnless',
+    'StreamFailFast',
     'StreamResult',
     'StreamSummary',
     'StreamToDict',
+    'TestControl',
     'ThreadsafeForwardingResult',
     'try_import',
     'try_imports',
@@ -73,11 +75,13 @@ else:
         CopyStreamResult,
         ExtendedToOriginalDecorator,
         MultiTestResult,
+        StreamFailFast,
         StreamResult,
         StreamSummary,
         StreamToDict,
         Tagger,
         TestByTestResult,
+        TestControl,
         TestResult,
         TestResultDecorator,
         TextTestResult,
diff --git a/testtools/testresult/__init__.py b/testtools/testresult/__init__.py
index 29ee39d..cc00e57 100644
--- a/testtools/testresult/__init__.py
+++ b/testtools/testresult/__init__.py
@@ -6,11 +6,13 @@ __all__ = [
     'CopyStreamResult',
     'ExtendedToOriginalDecorator',
     'MultiTestResult',
+    'StreamFailFast',
     'StreamResult',
     'StreamSummary',
     'StreamToDict',
     'Tagger',
     'TestByTestResult',
+    'TestControl',
     'TestResult',
     'TestResultDecorator',
     'TextTestResult',
@@ -21,11 +23,13 @@ from testtools.testresult.real import (
     CopyStreamResult,
     ExtendedToOriginalDecorator,
     MultiTestResult,
+    StreamFailFast,
     StreamResult,
     StreamSummary,
     StreamToDict,
     Tagger,
     TestByTestResult,
+    TestControl,
     TestResult,
     TestResultDecorator,
     TextTestResult,
diff --git a/testtools/testresult/real.py b/testtools/testresult/real.py
index c20b918..fde4a0d 100644
--- a/testtools/testresult/real.py
+++ b/testtools/testresult/real.py
@@ -6,10 +6,12 @@ __metaclass__ = type
 __all__ = [
     'ExtendedToOriginalDecorator',
     'MultiTestResult',
+    'StreamFailFast',
     'StreamResult',
     'StreamSummary',
     'StreamToDict',
     'Tagger',
+    'TestControl',
     'TestResult',
     'TestResultDecorator',
     'ThreadsafeForwardingResult',
@@ -400,6 +402,19 @@ class CopyStreamResult(StreamResult):
         domap(methodcaller('status', *args, **kwargs), self.targets)
 
 
+class StreamFailFast(StreamResult):
+    """Call the supplied callback if an error is seen in a stream."""
+
+    def __init__(self, callback):
+        self.callback = callback
+
+    def status(self, test_id=None, test_status=None, test_tags=None,
+        runnable=True, file_name=None, file_bytes=None, eof=False,
+        mime_type=None, route_code=None, timestamp=None):
+        if test_status in ('uxsuccess', 'fail'):
+            self.callback()
+
+
 class StreamToDict(StreamResult):
     """A specialised StreamResult that emits a callback as tests complete.
 
@@ -595,6 +610,23 @@ class StreamSummary(StreamToDict):
         self.unexpectedSuccesses.append(case)
 
 
+class TestControl(object):
+    """Controls a running test run, allowing it to be interrupted.
+    
+    :attribute shouldStop: If True, tests should no run and should instead
+        return without running. Similarly a TestSuite should abort running
+        and return immediately.
+    """
+
+    def __init__(self):
+        super(TestControl, self).__init__()
+        self.shouldStop = False
+
+    def stop(self):
+        """Indicate that tests should stop running."""
+        self.shouldStop = True
+
+
 class MultiTestResult(TestResult):
     """A test result that dispatches to many test results."""
 
diff --git a/testtools/tests/test_testresult.py b/testtools/tests/test_testresult.py
index f766840..1e137ad 100644
--- a/testtools/tests/test_testresult.py
+++ b/testtools/tests/test_testresult.py
@@ -22,11 +22,13 @@ from testtools import (
     ExtendedToOriginalDecorator,
     MultiTestResult,
     PlaceHolder,
+    StreamFailFast,
     StreamResult,
     StreamSummary,
     StreamToDict,
     Tagger,
     TestCase,
+    TestControl,
     TestResult,
     TestResultDecorator,
     TestByTestResult,
@@ -546,6 +548,12 @@ class TestStreamToDictContract(TestCase, TestStreamResultContract):
         return StreamToDict(lambda x:None)
 
 
+class TestStreamFailFastContract(TestCase, TestStreamResultContract):
+
+    def _make_result(self):
+        return StreamFailFast(lambda:None)
+
+
 class TestDoubleStreamResultEvents(TestCase):
 
     def test_startTestRun(self):
@@ -700,6 +708,47 @@ class TestStreamToDict(TestCase):
         self.assertEqual(["C", None], tests[1]['timestamps'])
 
 
+class TestStreamFailFast(TestCase):
+
+    def test_inprogress(self):
+        result = StreamFailFast(self.fail)
+        result.status('foo', 'inprogress')
+
+    def test_exists(self):
+        result = StreamFailFast(self.fail)
+        result.status('foo', 'exists')
+
+    def test_xfail(self):
+        result = StreamFailFast(self.fail)
+        result.status('foo', 'xfail')
+
+    def test_uxsuccess(self):
+        calls = []
+        def hook():
+            calls.append("called")
+        result = StreamFailFast(hook)
+        result.status('foo', 'uxsuccess')
+        result.status('foo', 'uxsuccess')
+        self.assertEqual(['called', 'called'], calls)
+
+    def test_success(self):
+        result = StreamFailFast(self.fail)
+        result.status('foo', 'success')
+
+    def test_fail(self):
+        calls = []
+        def hook():
+            calls.append("called")
+        result = StreamFailFast(hook)
+        result.status('foo', 'fail')
+        result.status('foo', 'fail')
+        self.assertEqual(['called', 'called'], calls)
+
+    def test_skip(self):
+        result = StreamFailFast(self.fail)
+        result.status('foo', 'skip')
+
+
 class TestStreamSummary(TestCase):
 
     def test_attributes(self):
@@ -847,6 +896,17 @@ testtools.matchers._impl.MismatchError: Differences: [
         self.assertEqual("foo.bar", result.unexpectedSuccesses[0].id())
 
 
+class TestTestControl(TestCase):
+
+    def test_default(self):
+        self.assertEqual(False, TestControl().shouldStop)
+
+    def test_stop(self):
+        control = TestControl()
+        control.stop()
+        self.assertEqual(True, control.shouldStop)
+
+
 class TestTestResult(TestCase):
     """Tests for 'TestResult'."""
